### PR TITLE
Implement GetObjectAttributes in mountpoint-s3-client

### PR DESCRIPTION
**Pull request by monthonk**
This PR implement [GetObjectAttributes](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObjectAttributes.html) for our S3CrtClient.

I leave the failure client unimplemented for now since some other operations are also not implemented, so I think we might want to refactor the failure client and implement them in another PR.

Closes https://github.com/awslabs/mountpoint-s3/issues/115.

---

By submitting this pull request,
I confirm that my contribution is made under the terms of the Apache 2.0 license
and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3-client/src/s3_crt_client/get_object_attributes.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    InvalidResponse(xmltree::Element, String),
+
+    #[error("XML parsing error: {0:?}")]
+    Xml(#[from] xmltree::ParseError),
+
+    #[error("Missing field {1} from XML element {0:?}")]
+    MissingField(xmltree::Element, String),
+
+    #[error("Failed to parse field {0} from string")]
+    FromStr(String),
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3-client/src/s3_crt_client/get_object_attributes.rs:28]: It's `InvalidResponse` or if we parse it to a wrong type.

File: mountpoint-s3-client/tests/get_object_attributes.rs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    let result = result.unwrap();
+    assert!(result.etag.is_some());
+    assert_eq!(result.storage_class, Some("STANDARD".to_owned()));
+    assert_eq!(result.object_size, Some(object_size));
+
+    assert!(result.checksum.is_some());
+    let checksum = result.checksum.unwrap();
+    assert!(checksum.checksum_crc32c.is_some());
+
+    assert!(result.object_parts.is_some());
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3-client/tests/get_object_attributes.rs]: Looks like `ObjectParts` is returned but not the inner `Parts`, added the test for it.

File: mountpoint-s3-client/src/object_client.rs [hunk 3]
```diff
@@ -60,6 +60,14 @@ pub trait ObjectClient {
         params: &PutObjectParams,
         contents: impl Stream<Item = impl AsRef<[u8]> + Send> + Send,
     ) -> ObjectClientResult<PutObjectResult, PutObjectError, Self::ClientError>;
+
+    /// Retrieves all the metadata from an object without returning the object contents.
+    async fn get_object_attributes(
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3-client/src/object_client.rs:65]: I will update this interface.

File: mountpoint-s3-client/src/s3_crt_client/get_object_attributes.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            })
+        }
+
+        let mut object_parts = None;
+        if let Some(mut object_parts_elem) = element.take_child("ObjectParts") {
+            let is_truncated = get_field(&object_parts_elem, "IsTruncated")?;
+            let max_parts = get_field(&object_parts_elem, "MaxParts")?;
+            let next_part_number_marker = get_field(&object_parts_elem, "NextPartNumberMarker")?;
+            let part_number_marker = get_field(&object_parts_elem, "PartNumberMarker")?;
+            let parts_count = get_field(&object_parts_elem, "PartsCount")?;
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3-client/src/s3_crt_client/get_object_attributes.rs]: Changed all of them to `Option`.

File: mountpoint-s3-client/src/s3_crt_client/get_object_attributes.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamesbornholt) [reply mountpoint-s3-client/src/s3_crt_client/get_object_attributes.rs:30]: I guess my point is: do you think we need both error types, or could all the `FromStr` errors just be `InvalidResponse` instead?

[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3-client/src/s3_crt_client/get_object_attributes.rs:30]: That's what I'm not sure so I created a new error type, but yes I think we can use `InvalidResponse`. I will switch to that one.

**Pull request merged**